simplecpp.cpp

```diff
@@ -2302,15 +2302,192 @@ static void simplifyName(simplecpp::TokenList &expr)
+
+        unsigned long long value;
+        
+        if(tok[pos++] == '\\') {
+            switch(char escape = tok[pos++]) {
+                case '\'':
+                case '"':
+                case '?':
+                case '\\':
+                    value = static_cast<unsigned char>(escape);
+                    break;
+
+                case 'a': value = static_cast<unsigned char>('\a'); break;
+                case 'b': value = static_cast<unsigned char>('\b'); break;
+                case 'f': value = static_cast<unsigned char>('\f'); break;
+                case 'n': value = static_cast<unsigned char>('\n'); break;
+                case 'r': value = static_cast<unsigned char>('\r'); break;
+                case 't': value = static_cast<unsigned char>('\t'); break;
+                case 'v': value = static_cast<unsigned char>('\v'); break;
+
+                // ESC extension
+                case 'e': value = static_cast<unsigned char>('\x1b'); break;
+
+                case '0':
+                case '1':
+                case '2':
+                case '3':
+                case '4':
+                case '5':
+                case '6':
+                case '7':
+                          // octal escape sequences consist of 1 to 3 digits
+                          value = stringToULLbounded(tok, --pos, 8, 1, 3);
+                          break;
+
+                case 'x':
+                          // hexadecimal escape sequences consist of at least 1 digit
+                          value = stringToULLbounded(tok, pos, 16);
+                          break;
+
+                case 'u':
+                case 'U': {
+                              // universal character names have exactly 4 or 8 digits
+                              std::size_t ndigits = (escape == 'u' ? 4 : 8);
+                              value = stringToULLbounded(tok, pos, 16, ndigits, ndigits);
+
+                              // UTF-8 encodes code points above 0x7f in multiple code units
+                              // code points above 0x10ffff are not allowed
+                              if(((narrow || utf8) && value > 0x7f) || (utf16 && value > 0xffff) || value > 0x10ffff)
+                                  throw std::runtime_error("code point too large");
+                              
+                              if(value >= 0xd800 && value <= 0xdfff)
+                                  throw std::runtime_error("surrogate code points not allowed in universal character names");
+
+                              break;
+                          }
+
+                default:
+                          throw std::runtime_error("ivalid escape sequence");
+            }
+        } else {
+            value = static_cast<unsigned char>(tok[pos-1]);
+
+            if(value > 0x7f)
+                throw std::runtime_error("unexpected byte in token");
```

Comment:
I limited the range only because I can't easily translate between character set encodings or the `\u` escape sequence matching a non-ASCII to the correct value, so `'å'` (code point `0xe5`) might work if the host, source and target character set encodings are assumed to match (probably latin-1), while `'\u00e5'` will not. (Although in this specific case the numeric value would just be `0xe5`.) I am fine with removing the check though to make non-ASCII source files at least work partially.

Reply:
Actually, I would like to keep the check at least for the non-narrow literal cases, since they are supposed to be UTF encoded and as I mentioned, I cannot translate them correctly, so they would generally have incorrect values without the check. For the `u8` literals it is always wrong and for the `u` and `U` literals it is only some going to be correct if the source encoding it latin-1.

Reply:
ok you can keeo the check for non-narrow literal cases if that is impossible UTF encoding.. I am not very good at what codes are possible..